gnucash/gnome-utils/gnc-autoclear.cpp

```diff
@@ -0,0 +1,271 @@
+ *                                                                  *
+ * This program is free software; you can redistribute it and/or    *
+ * modify it under the terms of the GNU General Public License as   *
+ * published by the Free Software Foundation; either version 2 of   *
+ * the License, or (at your option) any later version.              *
+ *                                                                  *
+ * This program is distributed in the hope that it will be useful,  *
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of   *
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the    *
+ * GNU General Public License for more details.                     *
+ *                                                                  *
+ * You should have received a copy of the GNU General Public License*
+ * along with this program; if not, contact:                        *
+ *                                                                  *
+ * Free Software Foundation           Voice:  +1-617-542-5942       *
+ * 51 Franklin Street, Fifth Floor    Fax:    +1-617-542-2652       *
+ * Boston, MA  02110-1301,  USA       gnu@gnu.org                   *
+ *******************************************************************/
+#include <config.h>
+
+#include <gtk/gtk.h>
+#include <glib/gi18n.h>
+#include <vector>
+#include <unordered_map>
+
+#include "inttypes.h"
+#include "Account.h"
+#include "Split.h"
+#include "Transaction.h"
+#include "gncOwner.h"
+#include "qof.h"
+#include "gnc-autoclear.h"
+
+static QofLogModule log_module = GNC_MOD_GUI;
+
+/* the following functions are used in window-autoclear: */
+
+typedef enum
+{
+    AUTOCLEAR_OVERLOAD = 1,
+    AUTOCLEAR_UNABLE,
+    AUTOCLEAR_MULTIPLE,
+    AUTOCLEAR_NOP,
+} autoclear_error_type;
+
+#define MAX_AUTOCLEAR_SECONDS 10
+
+#define MAX_AUTOCLEAR_TICKS CLOCKS_PER_SEC * MAX_AUTOCLEAR_SECONDS
+
+using SplitVec = std::vector<Split*>;
+
+struct WorkItem
+{
+    gnc_numeric reachable_amount;
+    SplitVec splits_vector;
+    WorkItem (const gnc_numeric amount, SplitVec&& splits)
+        : reachable_amount (amount), splits_vector(std::move(splits)){}
+};
+
+static void status_update (GtkLabel *label, gchar *status)
+{
+    if (!label) return;
+    gtk_label_set_text (label, status);
+    while (gtk_events_pending ())
+        g_main_context_iteration (nullptr, FALSE);
+}
+
+static void looping_update_status (GtkLabel *label, guint nc_progress,
+                                   guint nc_size, guint size)
+{
+    if (!label)
+        return;
+
+    auto text = g_strdup_printf ("%u/%u splits processed, %u combos",
+                                       nc_progress, nc_size, size);
+    status_update (label, text);
+    g_free (text);
+}
+
+class Hasher
+{
+public:
+    size_t operator() (gnc_numeric const& key) const
+    {
+        double d = gnc_numeric_to_double (key);
+        return g_double_hash (&d);
```

Comment:
Be *very* careful with `gnc_numeric_to_double`: It has a lot less precision than one would expect from the 64 bits it consumes. It will likely work just fine with OECD member currencies but may not for highly inflated developing ones at one end or highly deflated ones like Bitcoin.